libs/ml/include/ml/ops/loss_functions/mean_square_error_loss.hpp

```diff
@@ -32,24 +32,39 @@ template <class T>
 class MeanSquareErrorLoss : public Ops<T>
 {
 public:
-  using ArrayType     = T;
-  using DataType      = typename ArrayType::Type;
-  using SizeType      = typename ArrayType::SizeType;
+  using TensorType    = T;
+  using DataType      = typename TensorType::Type;
+  using SizeType      = typename TensorType::SizeType;
   using VecTensorType = typename Ops<T>::VecTensorType;
+  using SPType        = OpMeanSquareErrorSaveableParams<T>;
 
-  explicit MeanSquareErrorLoss(ArrayType const &weightings = ArrayType())
+  explicit MeanSquareErrorLoss(SPType const &sp)
+    : Ops<T>(sp)
+    , weightings_(sp.weightings)
+  {}
+
+  explicit MeanSquareErrorLoss(TensorType const &weightings = TensorType())
     : weightings_(weightings)
   {}
+
   ~MeanSquareErrorLoss() override = default;
 
-  void Forward(VecTensorType const &inputs, ArrayType &output) override
+  std::shared_ptr<SaveableParamsInterface> GetOpSaveableParams() override
+  {
+    auto ret = std::make_shared<SPType>();
+
+    ret->weightings = weightings_;
+    return ret;
+  }
+
+  void Forward(VecTensorType const &inputs, TensorType &output) override
   {
     assert(inputs.size() == 2);
     assert(inputs.at(0)->shape() == inputs.at(1)->shape());
 
     if (weightings_.size() == static_cast<SizeType>(0))
     {
-      output(0, 0) = fetch::math::MeanSquareError((*inputs.at(0)), (*inputs.at(1)));
+      output(0, 0) = fetch::math::MeanSquareError(*(inputs.at(0)), *(inputs.at(1)));
```

Comment:
I don't think that's correct in this case. Generally speaking the overhead of setting up an iterator is very small (instantiate class with ~5 or 6 integers, then 1 access call), but not zero. Here, calling operator(0, 0) makes probably 2 or 3 calls (you have two unroll the indices only twice); so that's not great but probably about equivalent or maybe even better.

In any case this is a very small optimisation detail that could be solved with benchmarks at a later date since the impact will be so low